[Spark-21842][Mesos] Support Kerberos ticket renewal and creation in Mesos

ArtRand opened the pull request:
## What changes were proposed in this pull request?
tl;dr: Add a class, `MesosHadoopDelegationTokenManager` that updates delegation tokens on a schedule on the behalf of Spark Drivers. Broadcast renewed credentials to the executors.

## The problem
We recently added Kerberos support to Mesos-based Spark jobs as well as Secrets support to the Mesos Dispatcher (SPARK-16742, SPARK-20812, respectively). However the delegation tokens have a defined expiration. This poses a problem for long running Spark jobs (e.g. Spark Streaming applications). YARN has a solution for this where a thread is scheduled to renew the tokens they reach 75% of their way to expiration. It then writes the tokens to HDFS for the executors to find (uses a monotonically increasing suffix). 

## This solution
We replace the current method in `CoarseGrainedSchedulerBackend` which used to discard the token renewal time with a protected method `fetchHadoopDelegationTokens`. Now the individual cluster backends are responsible for overriding this method to fetch and manage token renewal. The delegation tokens themselves, are still part of the `CoarseGrainedSchedulerBackend` as before. 
In the case of Mesos renewed Credentials are broadcasted to the executors. This maintains all transfer of Credentials within Spark (as opposed to Spark-to-HDFS). It also does not require any writing of Credentials to disk. It also does not require any GC of old files. 

## How was this patch tested?
Manually against a Kerberized HDFS cluster.


Thank you for the reviews.

Commit:
basic implementation of MesosCredentialRenewer

Commit:
pull upstream

AmplabJenkins commented:
Can one of the admins verify this patch?

skonto commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -174,6 +178,16 @@ private[spark] class CoarseGrainedExecutorBackend(
+    val creds = SparkHadoopUtil.get.deserialize(tokens)
+    // decode tokens and add them to the credentials
+    UserGroupInformation.getCurrentUser.addCredentials(SparkHadoopUtil.get.deserialize(tokens))
+  }
```
use the val above: UserGroupInformation.getCurrentUser.addCredentials(creds)

skonto commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -0,0 +1,150 @@
+    }
+    val tempCreds = ugi.getCredentials
+    val hadoopConf = SparkHadoopUtil.get.newConfiguration(conf)
+    var nextRenewalTime = Long.MaxValue
```
Same as spark.yarn.credentials.renewalTime, should not we have a common value somewhere?

skonto replied:
When driver is restarted in case of yarn the old renewalTime is restored: https://github.com/apache/spark/blob/e92ffe6f1771e3fe9ea2e62ba552c1b5cf255368/streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala#L59
Does the code here cover this?

ArtRand replied:
Right now when the `MesosCredentialRenewer` is initialized, it renews the current tokens and sets the renewal time to whatever the expiration time of those tokens is. On a driver restart, the same thing would happen. We could add `spark.yarn.credentials.renewalTime` as an override, but if the driver restarts, say 2 days later, `spark.yarn.credentials.renewalTime` is no longer relevant and it'll just immediately renew anyways. 

Relavent code: 
https://github.com/mesosphere/spark/blob/spark-21842-450-kerberos-ticket-renewal/resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala#L210 
^^ Where the initial renewal time is set
https://github.com/mesosphere/spark/blob/spark-21842-450-kerberos-ticket-renewal/resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala#L66
^^ where we initialize the renewal time if the renewal time has passed

skonto replied:
Ok so we always re-new when we start by fetching the tokens, got it.

skonto commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -123,6 +125,8 @@ private[spark] class CoarseGrainedExecutorBackend(
           executor.stop()
         }
       }.start()
+    case UpdateDelegationTokens(tokens) =>
```
Let's add a message that this is received only in mesos case, since CoarseGrainedExecutorBackend is used by both yarn and standalone.

skonto commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -174,6 +178,16 @@ private[spark] class CoarseGrainedExecutorBackend(
 
 private[spark] object CoarseGrainedExecutorBackend extends Logging {
 
+  private def addDelegationTokens(tokens: Array[Byte], sparkConf: SparkConf) {
```
Should we move this to SparkHadoopUtil and re-use 
methods such as: https://github.com/apache/spark/blob/d7b1fcf8f0a267322af0592b2cb31f1c8970fb16/core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala#L131 ?

skonto commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -159,6 +159,13 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+        for ((x, executorData) <- executorDataMap) {
+          executorData.executorEndpoint.send(UpdateDelegationTokens(tokens))
+        }
+
```
remove space

skonto commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -0,0 +1,150 @@
+    val hadoopConf = SparkHadoopUtil.get.newConfiguration(conf)
+    val creds = SparkHadoopUtil.get.deserialize(bytes)
+    val intervals = creds.getAllTokens.asScala.flatMap { t =>
+      Try {
```
t -> token
This method does not return an interval, it just returns the new expiration time. 
Compare with: https://github.com/apache/spark/blob/b9ab791a9efb0dc165ba283c91acf831fa6be5d8/core/src/main/scala/org/apache/spark/deploy/security/HadoopFSDelegationTokenProvider.scala#L102

skonto commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala:
```diff
@@ -194,6 +198,24 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
+            new MesosCredentialRenewer(
+              conf,
+              hadoopDelegationTokenManager.get,
+              MesosCredentialRenewer.getTokenRenewalInterval(hadoopDelegationCreds.get, conf),
```
hadoopDelegationCreds.get. Should we check against none? Creds are loaded in CoarseGrainedSchedulerBackend but if they are missing we should fail here?

skonto commented:
When executor is started it asks from CoarseSchedulerBackend the spark config which contains the hadoop credentials https://github.com/apache/spark/blob/7c92351f43ac4b1710e3c80c78f7978dad491ed2/core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala#L235.
As we have discussed this is not safe. Using rpc an arbitrary executor could register to the scheduler and get tokens. Does this code handle this, do we authenticate executors? How about encryption at least at the rpc level? The latter is not supported on mesos (spark.io.encryption.enabled).

susanxhuynh reviewed (changes requested):
Thanks, Art! Would you mind adding a note about broadcasting the tokens to executors in the PR description? Also, see comments.

skonto commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -159,6 +159,13 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+
+      case UpdateDelegationTokens(tokens) =>
+        logDebug("Asking each executor to update HDFS delegation tokens")
+        for ((x, executorData) <- executorDataMap) {
```
Alternatively executorDataMap.values.foreach(...)

Commit:
working through comments

Commit:
cleanup

Commit:
remove unused imports

ArtRand commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -0,0 +1,150 @@
+            broadcastDelegationTokens(creds)
+          } catch {
+            case e: Exception =>
+              // Log the error and try to write new tokens back in an hour
```
good catch, I changed the code to match the YARN equivalent.

susanxhuynh commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala:
```diff
@@ -198,16 +198,19 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
       val credentialRenewerThread = new Thread {
         setName("MesosCredentialRenewer")
         override def run(): Unit = {
+          val dummy: Option[Array[Byte]] = None
```
What is this for?

ArtRand replied:
whoops!

susanxhuynh commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -63,7 +63,8 @@ class MesosCredentialRenewer(
     def scheduleRenewal(runnable: Runnable): Unit = {
-      val remainingTime = timeOfNextRenewal - System.currentTimeMillis()
+      // val remainingTime = timeOfNextRenewal - System.currentTimeMillis()
+      val remainingTime = 5000
```
Why 5000?

ArtRand replied:
well that's embarrassing, just a debugging tool that I forgot to remove.

Commit:
remove debugging code

susanxhuynh reviewed (approved):
LGTM

kalvinnchau commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -0,0 +1,150 @@
+    logInfo(s"Attempting to login with ${conf.get("spark.yarn.principal", null)}")
+    // Get new delegation tokens by logging in with a new UGI
+    // inspired by AMCredentialRenewer.scala:L174
+    val ugi = if (mode == "keytab") {
```
I don't see where it refreshes the delegation tokens until the max-lifetime, then re-login with the keytab to get a new delegation tokens that'll last until the max-lifetime.

Does this skip over the potential issues with expiring delegation tokens (after the max-lifetime, 7 days default) by just re-logging in with the keytab every time the delegation tokens need to refresh, and then grabbing a new set of delegation tokens?

ArtRand replied:
Hello @kalvinnchau You are correct, all this does is keep track of when the tokens will expire and renew them at that time. Part of my motivation for doing this is to avoid writing any files to disk (like new TGTs, if that's what you're suggesting). We can simply mount the keytab via the Mesos secrets primitive, then renew the tokens every so often. In order to be consistent I tried to keep this solution as close to YARN as possible.

vanzin replied:
The correct way would be for the credential management code to differentiate between token creation and token renewal; that way it would renew tokens at the renewal internal and create new ones after the max lifetime.

But it seems the original implementation took a shortcut and just creates new one instead of renewing existing ones; changing that would require changes in the credential provider interfaces, so this is enough for now.

kalvinnchau commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala:
```diff
@@ -194,6 +198,26 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
+            new MesosCredentialRenewer(
+              conf,
+              hadoopDelegationTokenManager.get,
+              MesosCredentialRenewer.getTokenRenewalTime(hadoopDelegationCreds.get, conf),
```
This sets the first renewal time to be the expiration time of the token.

It should be similar to the way next renewal time in the MesosCredentialRenewer class is calculated so that it renews the first token after 75% of expiration time has passed:

```scala
val currTime = System.currentTimeMillis()
val renewTime = MesosCredentialRenewer.getTokenRenewalTime(hadoopDelegationCreds.get, conf)
val rt = 0.75 * (renewTime - currTime)

val credentialRenewer =
   new MesosCredentialRenewer(
     conf,
     hadoopDelegationTokenManager.get,
     (currTime + rt).toLong,
     driverEndpoint)
credentialRenewer.scheduleTokenRenewal()
```

Commit:
use 75 percent of renewal time for the first renewal also

ArtRand commented:
Hey @kalvinnchau good catch on the first renewal time. I believe I addressed it. Have a look. Thanks again.

kalvinnchau commented:
@ArtRand thanks! I've been testing a local version of doing that, I'll pull that change in and test it as well.

kalvinnchau commented:
@ArtRand 
I'm running into an issue where it seems like the delegation tokens are being sent to the executors (as in I see the logs stating that the tokens are being broadcast), but the old delegation tokens are still in use.

When the job first starts up it created token `777222`, then ~18 hours in the refresh occurs:
`[INFO ] 2017-09-26 17:59:20.639 [Credential Refresh Thread-0] DFSClient - Created HDFS_DELEGATION_TOKEN token 772826 for <principal> on ha-hdfs:<namenode>`


`[INFO ] 2017-09-26 17:59:20.747 [Credential Refresh Thread-0] MesosCredentialRenewer - Sending new tokens to all executors
`

Then at ~24 hour mark I get the exception:
`
Caused by: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.security.token.SecretManager$InvalidToken): token (HDFS_DELEGATION_TOKEN token 772222 for <principal>) is expired
`

Are you aware of anything that needs to be done to tell the thread to use the newer tokens? Or do the older tokens need to be removed from the UGI?

ArtRand commented:
Hey @kalvinnchau thanks for having the patience to try this. This is a curious error though. 

If you look at the `addAll` method called by `UserGroupInformation.addCredentials()` it should overwrite the current credentials. 

I tried to reproduce your error, but being less patient, I changed my HDFS setup to request the tokens be updated every minute instead of everyday by adding the following to hdfs-site.xml:
```
    <property>
        <name>dfs.namenode.delegation.token.max-lifetime</name>
        <value>60000</value>
    </property>
```

I added some logging to the executor backend to check if they were indeed being updated. 

```
case UpdateDelegationTokens(tokens) =>
      logInfo("Got request to update tokens")
      val oldCreds = UserGroupInformation.getCurrentUser.getCredentials
      for (t <- oldCreds.getAllTokens.asScala) {
        logInfo(s"Old Creds ${DelegationTokenIdentifier.stringifyToken(t)}")
      }
      val creds = SparkHadoopUtil.get.deserialize(tokens)
      for (t <- creds.getAllTokens.asScala) {
        val s = DelegationTokenIdentifier.stringifyToken(t)
        logInfo(s"Got new tokens $s")
      }
      SparkHadoopUtil.get.addDelegationTokens(tokens, env.conf)
      val newCreds = UserGroupInformation.getCurrentUser.getCredentials
      for (t <- newCreds.getAllTokens.asScala) {
        logInfo(s"New creds ${DelegationTokenIdentifier.stringifyToken(t)}")
      }
```

and indeed when I check the logs, indeed the token number has been updated. 
```
17/09/28 03:32:58 INFO CoarseGrainedExecutorBackend: Got request to update tokens
17/09/28 03:32:58 INFO CoarseGrainedExecutorBackend: Old Creds HDFS_DELEGATION_TOKEN token 29 for hdfs on ha-hdfs:hdfs
17/09/28 03:32:59 INFO CoarseGrainedExecutorBackend: Got new tokens HDFS_DELEGATION_TOKEN token 31 for hdfs on ha-hdfs:hdfs
17/09/28 03:32:59 INFO CoarseGrainedExecutorBackend: New creds HDFS_DELEGATION_TOKEN token 31 for hdfs on ha-hdfs:hdfs
```
then some time later (in fact there was another update in the middle):
```
17/09/28 03:35:14 INFO CoarseGrainedExecutorBackend: Got request to update tokens
17/09/28 03:35:14 INFO CoarseGrainedExecutorBackend: Old Creds HDFS_DELEGATION_TOKEN token 34 for hdfs on ha-hdfs:hdfs
17/09/28 03:35:14 INFO CoarseGrainedExecutorBackend: Got new tokens HDFS_DELEGATION_TOKEN token 35 for hdfs on ha-hdfs:hdfs
17/09/28 03:35:14 INFO CoarseGrainedExecutorBackend: New creds HDFS_DELEGATION_TOKEN token 35 for hdfs on ha-hdfs:hdfs
```

I will run a 24h experiment to verify, but hopefully there is a way to validate that the update is working without waiting that long just to debug!

@vanzin Could you eyeball this, am I missing something obvious?

vanzin commented:
What you wrote sounds correct. However, I've seen errors like the ones above in the past, but haven't been able to fully debug them due to lack of logs. Part of it is because the user running the app didn't provide me the full logs; but also Spark currently doesn't have the logs you added in your debug code, not even at debug level, so we have only partial information about exactly what's going on from existing logs. (And these error tend to happen only after x days, so it's kind of a pain to reproduce.)

That all being said, though, as I mentioned, what you wrote sounds correct AFAIK.

vanzin commented:
Another thing to keep in mind is that different Hadoop versions have different bugs in this area, so if you use a version of the Hadoop client library that suffers from some issue, or are talking to an HDFS version that has some bug, that can cause problems.

Without researching more I only remember an issue that affects HA mode (HDFS-9276), but there might be others.

ArtRand commented:
Hello @vanzin 

Thanks for taking a look at this. Good to know that there can be downstream errors depending on the situation. 

Would very much appreciate a proper review on this work when you have some time, very keen on getting this into the next release.

kalvinnchau commented:
@ArtRand curious, what version of hadoop are you build spark against and what version is the cluster that you're running?

ArtRand commented:
@kalvinnchau I'm running Hadoop 2.6 on a DC/OS cluster with Mesos 1.4.0

vanzin commented:
ok to test

vanzin commented:
add to whitelist

SparkQA commented:
**[Test build #82602 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82602/testReport)** for PR 19272 at commit [`f5925fd`](https://github.com/apache/spark/commit/f5925fd2ab31c1d6825fa546102172ed8a55ae8b).

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -0,0 +1,154 @@
+    val keytab64 = conf.get("spark.yarn.keytab", null)
+    val tgt64 = System.getenv("KRB5CCNAME")
+    require(keytab64 != null || tgt64 != null, "keytab or tgt required")
+    require(keytab64 == null || tgt64 == null, "keytab and tgt cannot be used at the same time")
```
`KRB5CCNAME` is something that people might have in their environment for various reasons, so I'd avoid this requirement.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala:
```diff
@@ -194,6 +198,27 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
+      val credentialRenewerThread = new Thread {
+        setName("MesosCredentialRenewer")
+        override def run(): Unit = {
+          val rt = MesosCredentialRenewer.getTokenRenewalTime(hadoopDelegationCreds.get, conf)
```
So, you need this because `hadoopDelegationCreds` doesn't keep the information about when the tokens should be renewed (a.k.a. the return value of `obtainDelegationTokens`). Perhaps some minor refactoring would help clean this up?

In fact, `hadoopDelegationCreds` is a `val`, so any executors that start after the initial token set expires will fail, no? Because they'll fetch `hadoopDelegationCreds` from the driver, and won't get the `UpdateDelegationTokens` until it's way too late.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -0,0 +1,154 @@
+    conf: SparkConf,
+    tokenManager: HadoopDelegationTokenManager,
+    nextRenewal: Long,
+    de: RpcEndpointRef) extends Logging {
```
How about a more descriptive variable name?

ArtRand replied:
fixed.

SparkQA commented:
**[Test build #82602 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82602/testReport)** for PR 19272 at commit [`f5925fd`](https://github.com/apache/spark/commit/f5925fd2ab31c1d6825fa546102172ed8a55ae8b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82602/
Test PASSed.

ArtRand commented:
@vanzin Thanks for the review. I'll address the comments ASAP.

Commit:
addressed comments, still pending testing

SparkQA commented:
**[Test build #82777 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82777/testReport)** for PR 19272 at commit [`3f22efe`](https://github.com/apache/spark/commit/3f22efe65e5083722dead8f2f3729ed54b8dd530).

SparkQA commented:
**[Test build #82777 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82777/testReport)** for PR 19272 at commit [`3f22efe`](https://github.com/apache/spark/commit/3f22efe65e5083722dead8f2f3729ed54b8dd530).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82777/
Test PASSed.

Commit:
final comments

SparkQA commented:
**[Test build #82870 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82870/testReport)** for PR 19272 at commit [`e522150`](https://github.com/apache/spark/commit/e522150c32ffca9f116197f80530554d3cb68b6c).

ArtRand commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -0,0 +1,154 @@
+    nextRenewal: Long,
+    de: RpcEndpointRef) extends Logging {
+  private val credentialRenewerThread =
+    Executors.newSingleThreadScheduledExecutor(
```
I also changed `AMCredentialRenewer` to the same.

ArtRand commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala:
```diff
@@ -194,6 +198,27 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
+      }
+
+      credentialRenewerThread.start()
+      credentialRenewerThread.join()
```
Yes, I believe so. If you look here (https://github.com/apache/spark/blob/6735433cde44b3430fb44edfff58ef8149c66c13/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala#L271) it's the same pattern. The thread needs to run as long as the application driver, correct?

vanzin replied:
There's a comment explaining why that thread exists right above the code you linked. Did you look at it?

Also, you're calling `join()` on the thread, so it's obviously going away.

ArtRand replied:
Yes, sorry, for some reason I understood you you to mean the credential renewer itself. I added a comment to the same effect as the YARN analogue.

vanzin replied:
I don't think you really understood why the YARN code needs a thread and why I'm telling you this code does not. Read the comment you added here again; what makes you think the current thread does not have access to those classes?

ArtRand replied:
Ok, you're probably right. It appears that the YARN code uses `setContextClassLoader(userClassLoader)` whereas in Mesos does not has a notion of `userClassLoader`. Therefore we don't need the separate thread in the Mesos code. Do I have this correct? Thanks for showing me this!

vanzin replied:
That's the gist of it.

SparkQA commented:
**[Test build #82870 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82870/testReport)** for PR 19272 at commit [`e522150`](https://github.com/apache/spark/commit/e522150c32ffca9f116197f80530554d3cb68b6c).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82870/
Test PASSed.

Commit:
add comments

SparkQA commented:
**[Test build #82919 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82919/testReport)** for PR 19272 at commit [`837157d`](https://github.com/apache/spark/commit/837157d1c76d1b9488025e8c294eb2839aeee5e3).

SparkQA commented:
**[Test build #82919 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82919/testReport)** for PR 19272 at commit [`837157d`](https://github.com/apache/spark/commit/837157d1c76d1b9488025e8c294eb2839aeee5e3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82919/
Test PASSed.

Commit:
remove extra thread in MesosCredentialRenewer

SparkQA commented:
**[Test build #82941 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82941/testReport)** for PR 19272 at commit [`c95f80b`](https://github.com/apache/spark/commit/c95f80b23d47ea4640cea2b4a185fa4bf9e9f33d).

SparkQA commented:
**[Test build #82941 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82941/testReport)** for PR 19272 at commit [`c95f80b`](https://github.com/apache/spark/commit/c95f80b23d47ea4640cea2b4a185fa4bf9e9f33d).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82941/
Test FAILed.

ArtRand commented:
Retest this please.

SparkQA commented:
**[Test build #82955 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82955/testReport)** for PR 19272 at commit [`c95f80b`](https://github.com/apache/spark/commit/c95f80b23d47ea4640cea2b4a185fa4bf9e9f33d).

SparkQA commented:
**[Test build #82955 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82955/testReport)** for PR 19272 at commit [`c95f80b`](https://github.com/apache/spark/commit/c95f80b23d47ea4640cea2b4a185fa4bf9e9f33d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82955/
Test PASSed.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterManager.scala:
```diff
@@ -17,7 +17,7 @@
 package org.apache.spark.scheduler.cluster.mesos
 
-import org.apache.spark.{SparkContext, SparkException}
+import org.apache.spark.SparkContext
```
Change not needed?

ArtRand replied:
`SparkException` is unused, not sure why it was there in the first place

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -159,6 +159,12 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+        // Update the driver's delegation tokens in case new executors are added later.
+        currentHadoopDelegationTokens = Some(tokens)
+        executorDataMap.values.foreach { ed =>
+          ed.executorEndpoint.send(UpdateDelegationTokens(tokens)) }
```
`}` goes in next line.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala:
```diff
@@ -213,6 +216,24 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
+      // The renewal time is ignored when getting the initial delegation tokens
+      // (CoarseGrainedSchedulerBackend.scala:getHadoopDelegationCreds), so we get the renewal
+      // time here and schedule a thread to renew them.
+      val renewalTime =
```
I still don't like this. You should not need to implement this separate method of getting the renewal time just because the current code is throwing out that information. Instead you should fix the code so that the information is preserved.

`getHadoopDelegationCreds` is called in only one place, so my suggestion would be to encapsulate initializing the token manager and getting the initial set of tokens into a single method (instead of the current two).

Then in that method's implementation you can get the initial set of tokens, initialize the renewer thread with the correct renewal period, and return the data needed by the scheduler.

ArtRand replied:
Ok, I agree. I refrained from touching the current code only because it was added it so recently. I assumed the omission of renewal time was intentional (or at least semi-intentional). I'll submit a refactor. Are you suggesting that we make the renewal thread part of `CoarseGrainedSchedulerBackend` (with the appropriate override for YARN/Mesos)?  Or make a resource-manager specific method?

vanzin replied:
It makes more sense to have it in `CoarseGrainedSchedulerBackend` (then YARN and Mesos can easily share it); my worry is exposing the functionality to standalone (it would work, but not be secure, which is a problem).

So maybe add a second method `def supportsTokenRenewal` that subclasses can override to return `true` or something.

ArtRand replied:
Maybe an overly simplistic solution, but what about putting this functionality into `HadoopDelegationTokenManager`?

vanzin replied:
I guess that would be ok. I was a little worried about mixing the two but they are kinda closely related...

vanzin replied:
But do note that you still can't start the token renewer automatically, since YARN still has its own renewer and things would get mixed up.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -0,0 +1,169 @@
+    if (renewalTimes.isEmpty) Long.MaxValue else renewalTimes.min
+  }
+
+  def getNextRenewalTime(t: Long): Long = {
```
This looks similar but not exactly the same as the logic in YARN's `writeNewCredentialsToHDFS`. Both should be using the same logic to calculate these things, so probably time for some refactoring.

vanzin commented on resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMCredentialRenewer.scala:
```diff
@@ -59,8 +59,7 @@ private[yarn] class AMCredentialRenewer(
   private val credentialRenewer =
-    Executors.newSingleThreadScheduledExecutor(
-      ThreadUtils.namedThreadFactory("Credential Refresh Thread"))
+    ThreadUtils.newDaemonSingleThreadScheduledExecutor("Credential Refresh Thread")
```
Normally you should avoid making changes that are not related to your PR.

ArtRand replied:
I generally agree, but I did this mostly to keep these two classes as congruent as possible. I can change it back if you feel strongly.

vanzin replied:
It's ok since you've done it, I was mostly making a generic comment.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -0,0 +1,169 @@
+    require(keytab != null || tgt != null, "A keytab or TGT required.")
+    // if both Keytab and TGT are detected we use the Keytab.
+    val (secretFile, mode) = if (keytab != null && tgt != null) {
+      logWarning(s"Keytab and TGT were detected, using keytab, unset $keytab to use TGT")
```
`${KEYTAB.key}`?

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -0,0 +1,169 @@
+  }
+
+  private def broadcastDelegationTokens(tokens: Array[Byte]): Unit = {
+    // send token to existing executors
```
Comment is redundant.

Commit:
refactor credential renewer, addressed comments

Commit:
merge upstream

Commit:
resolve import problem

SparkQA commented:
**[Test build #83235 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83235/testReport)** for PR 19272 at commit [`864ab7e`](https://github.com/apache/spark/commit/864ab7ec659a5071e0ed1a87d2448c507b815a79).

SparkQA commented:
**[Test build #83235 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83235/testReport)** for PR 19272 at commit [`864ab7e`](https://github.com/apache/spark/commit/864ab7ec659a5071e0ed1a87d2448c507b815a79).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83235/
Test PASSed.

Commit:
cleanup

ArtRand commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -233,10 +249,15 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
           sparkProperties,
           SparkEnv.get.securityManager.getIOEncryptionKey(),
-          hadoopDelegationCreds)
+          tokens)
```
could make `SparkAppConfig` just take a `RenewableDelegationTokens` object?

Commit:
more cleanup

SparkQA commented:
**[Test build #83262 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83262/testReport)** for PR 19272 at commit [`f93c551`](https://github.com/apache/spark/commit/f93c5513178628d6e7c33bd6cc941407484299fb).

SparkQA commented:
**[Test build #83262 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83262/testReport)** for PR 19272 at commit [`f93c551`](https://github.com/apache/spark/commit/f93c5513178628d6e7c33bd6cc941407484299fb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83262/
Test PASSed.

ArtRand commented:
@vanzin Thanks for the continued review. I addressed the comments you left. PTAL.

It seems simplest and more robust to keep the serialized `Credentials` within `CoarseGrainedSchedulerBackend`. The change is that now the renewal time is also kept along side. I tried to move as much duplicate code to `HadoopDelegationTokenManager` which remains a "toolbox" as before and doesn't retain much state. The abstract class `HadoopCredentialRenewer` is now part of the scheduler backend also. I refrained from changing too much of the YARN specific code.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/security/HadoopCredentialRenewer.scala:
```diff
@@ -0,0 +1,57 @@
+ * (and the dispersal of tokens) is resource-manager-specific, see implementations of this
+ * class for details.
+ */
+abstract class HadoopCredentialRenewer extends Logging {
```
`private[spark]`

vanzin commented on core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala:
```diff
@@ -55,7 +56,9 @@ private[spark] class HadoopDelegationTokenManager(
     s"${delegationTokenProviders.keys.mkString(", ")}.")
 
-  /** Construct a [[HadoopDelegationTokenManager]] for the default Hadoop filesystem */
+  /** Construct a [[HadoopDelegationTokenManager]] for the default Hadoop filesystem with a
```
Comment starts on next line. But the comment doesn't really seem accurate anyway.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala:
```diff
@@ -125,3 +141,5 @@ private[spark] class HadoopDelegationTokenManager(
   }
 }
+
+case class RenewableDelegationTokens(credentials: Array[Byte], nextRenewalTime: Long)
```
This is not a great name from this class; it doesn't provide any functionality related to renewing the tokens, so it's not really clear why it's called "renewable".

If this actually wrapped the `HadoopDelegationTokenManager` and provided a `renew` method or something, that would be better. But at the same time this is starting to make this API too confusing.

So two options here: either use the existing API and make the caller hold on to the two pieces of info (tokens and renewal time), or change the existing API. But don't add a new API that basically does the same thing.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala:
```diff
@@ -75,6 +78,17 @@ private[spark] class HadoopDelegationTokenManager(
       .toMap
   }
 
+  def getRenewableDelegationTokens(): Option[RenewableDelegationTokens] = {
```
I'm not too fond of the idea of having two methods for creating tokens (this and `obtainDelegationTokens`). One should be enough.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -102,8 +99,15 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
 
-  // Hadoop delegation tokens to be sent to the executors.
-  val hadoopDelegationCreds: Option[Array[Byte]] = getHadoopDelegationCreds()
+  def hadoopCredentialRenewer: Option[HadoopCredentialRenewer] = None
```
Why is this needed here? This class doesn't use it at all.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -102,8 +99,15 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+  def hadoopCredentialRenewer: Option[HadoopCredentialRenewer] = None
+
+  // Hadoop delegation tokens to be sent to the executors, can be updated as necessary.
+  var renewableDelegationTokens: Option[RenewableDelegationTokens] =
```
This is another part where the interface is brittle and a bit confusing. This should really be a private field, because there should be a single way for sub-classes to update it (send a `UpdateDelegationTokens` message). But it's not because Mesos needs to poke at it because the rest of the API is all a bit weird.

If instead you follow my original suggestion (have a abstract `initializeDelegationTokens()` method that Mesos overrides), then everything becomes a lot more localized and cleaner.

vanzin reviewed (commented):
Really not a fan of the new API changes. We should be making things simpler, not adding more specialized APIs that do the exact same thing, but in a slightly different way.

Commit:
move hadoop delegation token management into Mesos renewer

SparkQA commented:
**[Test build #83468 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83468/testReport)** for PR 19272 at commit [`b2fbcf2`](https://github.com/apache/spark/commit/b2fbcf250e9d5337ba87c6a3ea365831dd867fe6).

ArtRand commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -99,11 +96,8 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
   @volatile protected var currentExecutorIdCounter = 0
 
-  // hadoop token manager used by some sub-classes (e.g. Mesos)
-  def hadoopDelegationTokenManager: Option[HadoopDelegationTokenManager] = None
```
No longer needed because resource-manager backends (may) implement their own `initializeHadoopDelegationTokens`.

ArtRand commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -686,18 +687,7 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
     true
   }
 
-  protected def getHadoopDelegationCreds(): Option[Array[Byte]] = {
```
This method was only called once, and would discard the renewal time information limiting it's utility.

ArtRand commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala:
```diff
@@ -58,8 +62,9 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
 
-  override def hadoopDelegationTokenManager: Option[HadoopDelegationTokenManager] =
-    Some(new HadoopDelegationTokenManager(sc.conf, sc.hadoopConfiguration))
+  private lazy val hadoopCredentialRenewer: MesosCredentialRenewer =
```
For Mesos the credential renewer contains the tokens, the renewal time, and all logic to renew the tokens. Should never be evaluated (and tokens never initialized) if `UserGroupInformation.isSecurityEnabled` evaluates to `false`

SparkQA commented:
**[Test build #83468 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83468/testReport)** for PR 19272 at commit [`b2fbcf2`](https://github.com/apache/spark/commit/b2fbcf250e9d5337ba87c6a3ea365831dd867fe6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83468/
Test PASSed.

Commit:
remove unused imports

SparkQA commented:
**[Test build #83493 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83493/testReport)** for PR 19272 at commit [`4558cea`](https://github.com/apache/spark/commit/4558ceaee016361ee75eebb6a958fb98637060c3).

SparkQA commented:
**[Test build #83493 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83493/testReport)** for PR 19272 at commit [`4558cea`](https://github.com/apache/spark/commit/4558ceaee016361ee75eebb6a958fb98637060c3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83493/
Test PASSed.

ArtRand commented:
Hello @vanzin thanks for continuing to help with this. Please take another look at this refactor.

In this change, there is one place to interact with `hadoopDelegationTokens` from `CoarseGrainedSchedulerBackedn`: `initializeHadoopDelegationTokens`. This method contains the logic for initializing the tokens and setting a token renewer. It's also now resource-manager specific. This seems cleaner than having a `HadoopDelegationTokenManager` in `CoarseGrainedSchedulerBackend` because any "token management" will always want to wrap`HadoopDelegationTokenManager` so you can keep all the necessary information in one place. Of course, happy to discuss further.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala:
```diff
@@ -55,7 +55,9 @@ private[spark] class HadoopDelegationTokenManager(
     s"${delegationTokenProviders.keys.mkString(", ")}.")
 
-  /** Construct a [[HadoopDelegationTokenManager]] for the default Hadoop filesystem */
+  /**
```
This is not really changing anything, so I'd just revert changes to this file. Or, if you really want to, just keep the new `@param`s you're adding below.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -99,11 +96,8 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
-  // Hadoop delegation tokens to be sent to the executors.
-  val hadoopDelegationCreds: Option[Array[Byte]] = getHadoopDelegationCreds()
+  // Hadoop delegation tokens to be sent to the executors, can be updated as necessary.
+  protected var hadoopDelegationTokens: Option[Array[Byte]] = initializeHadoopDelegationTokens()
```
Why is this protected? There's no reason I can see for subclasses to need access to this field.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala:
```diff
@@ -213,6 +216,14 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
 
+    // check that the credentials are defined, even though it's likely that auth would have failed
+    // already if you've made it this far, then start the token renewer
+    if (hadoopDelegationTokens.isDefined) {
```
You shouldn't do this here, otherwise you need to keep that field `protected` in the parent class and that adds unnecessary coupling. Instead, do this in `initializeHadoopDelegationTokens`.

ArtRand replied:
I agree that I shouldn't need to use the conditional `hadoopDelegationTokens.isDefined`, however there will need to be some check (`UserGroupInformation.isSecurityEnabled` or similar) to pass the `driverEndpoint` to the renewer/manager here. When the initial tokens are generated `driverEndpoint` is still `None` because `start()` hasn't been called yet.

ArtRand replied:
I may have spoke too soon, there might be a way..

vanzin replied:
You could call `initializeHadoopDelegationTokens` in `start` after everything that's needed is initialized. It would also better follow the scheduler's lifecycle.

ArtRand replied:
Check out the patch now. `hadoopDelegationTokens` now calls `initializeHadoopDelegationTokens` (renamed `fetchHadoopDelegationTokens`) by name:
```scala
  private val hadoopDelegationTokens: () => Option[Array[Byte]] = fetchHadoopDelegationTokens
```
 This has the effect of only generating the first set of delegation tokens once the first `RetrieveSparkAppConfig` message is received. At this point, everything has been initialized because renewer (renamed `MesosHadoopDelegationTokenManager`) is evaluated lazily with the correct `driverEndpoint`. 

It's a bit confusing to just avoid an extra conditional. WDYT?

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -0,0 +1,166 @@
+ * serialized) are broadcast to all running executors. On the executor side, when new Tokens are
+ * recieved they overwrite the current credentials.
+ */
+class MesosCredentialRenewer(
```
`private[spark]`

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -0,0 +1,166 @@
+
+  private val (secretFile, mode) = getSecretFile(conf)
+
+  var (tokens: Array[Byte], timeOfNextRenewal: Long) = {
```
`private`?

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala:
```diff
@@ -0,0 +1,166 @@
+    (secretFile, mode)
+  }
+
+  def scheduleTokenRenewal(driverEndpoint: RpcEndpointRef): Unit = {
```
Why isn't this done in the constructor? There's a single call to this method, and the renewal interval could very easily be turned into a constructor arg.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala:
```diff
@@ -58,8 +60,9 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
-    Some(new HadoopDelegationTokenManager(sc.conf, sc.hadoopConfiguration))
+  private lazy val hadoopCredentialRenewer: MesosCredentialRenewer =
+    new MesosCredentialRenewer(
+      conf, new HadoopDelegationTokenManager(sc.conf, sc.hadoopConfiguration))
```
Why pass in a `HadoopDelegationTokenManager` if it's not used by this class? The renewer can create one itself.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala:
```diff
@@ -772,6 +783,14 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
+
+  override def initializeHadoopDelegationTokens(): Option[Array[Byte]] = {
+    if (UserGroupInformation.isSecurityEnabled) {
+      Some(hadoopCredentialRenewer.tokens)
```
So, seems to me that your "renewer" is doing more than just renewing tokens; it's also being used to generate the initial set. So aside from my comments about initializing the renewer here, you should also probably make this API a little cleaner. Right now there's too much coupling.

The renewer should to renewals only, otherwise it should be called something different.

vanzin reviewed (commented):
I still don't like the API. There are just too many touch points between the different classes in your patch, and non-trivial initialization order requirements. That makes the code brittle and hard to modify later.

Commit:
no more conditionals, start renewer automatically

SparkQA commented:
**[Test build #83607 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83607/testReport)** for PR 19272 at commit [`5f254e5`](https://github.com/apache/spark/commit/5f254e5369238903cce0ac012f8f7ff48ea70afb).

SparkQA commented:
**[Test build #83607 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83607/testReport)** for PR 19272 at commit [`5f254e5`](https://github.com/apache/spark/commit/5f254e5369238903cce0ac012f8f7ff48ea70afb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83607/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -123,6 +123,9 @@ private[spark] class CoarseGrainedExecutorBackend(
       }.start()
+
+    case UpdateDelegationTokens(tokenBytes) =>
+      SparkHadoopUtil.get.addDelegationTokens(tokenBytes, env.conf)
```
Can you add a `logInfo` saying the tokens are being updated? This has always been helpful when debugging issues with this feature on YARN.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -159,6 +151,12 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
         }
+
+      case UpdateDelegationTokens(newDelegationTokens) =>
+        // Update the driver's delegation tokens in case new executors are added later.
```
Stale comment?

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -236,7 +234,7 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
           sparkProperties,
           SparkEnv.get.securityManager.getIOEncryptionKey(),
-          hadoopDelegationCreds)
+          hadoopDelegationTokens.apply())
```
Can't you just call `fetchHadoopDelegationTokens()` directly?

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala:
```diff
@@ -0,0 +1,180 @@
+ * received they overwrite the current credentials.
+ */
+private[spark] class MesosHadoopDelegationTokenManager(
+    conf: SparkConf, hadoopConfig: Configuration,
```
One arg per line.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    if (principal == null) {
+      logInfo(s"Using mode: $mode to retrieve Hadoop delegation tokens")
```
You should probably assert that mode is "tgt" in this case.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala:
```diff
@@ -0,0 +1,180 @@
+    scheduleRenewal(credentialRenewerRunnable)
+  }
+
+  private def getNewDelegationTokens: Array[Byte] = {
```
Prefer to use `()` in methods that do non-trivial things.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala:
```diff
@@ -0,0 +1,180 @@
+    } catch {
+      case e: Exception =>
+        throw new IllegalStateException("Failed to initialize Hadoop delegation tokens\n" +
+          s"\tPricipal: $principal\n\tmode: $mode\n\tsecret file $secretFile\n\tException: $e")
```
Use `e` as the cause of the exception you're throwing.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala:
```diff
@@ -0,0 +1,180 @@
+  private def getSecretFile(conf: SparkConf): (String, String) = {
+    val keytab = conf.get(config.KEYTAB).orNull
+    val tgt = conf.getenv("KRB5CCNAME")
+    require(keytab != null || tgt != null, "A keytab or TGT required.")
```
Is that really the case? `KRB5CCNAME` is not a required env variable. It has a default value, and the `UGI` class will use the credentials from the default location if they're available (and reloading the cache periodically).

So I think you don't really need this, but just to track whether there's a principal and keytab. And you don't need to call `getUGIFromTicketCache` later on since I'm pretty sure UGI takes care of that for you.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala:
```diff
@@ -0,0 +1,180 @@
+
+    val tempCreds = ugi.getCredentials
+    val hadoopConf = SparkHadoopUtil.get.newConfiguration(conf)
+    var nextRenewalTime = Long.MaxValue
```
```
val nextRenewalTime = ugi.doAs(new PrivilegedExceptionAction[Long] { ... }
```

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala:
```diff
@@ -0,0 +1,180 @@
+  }
+
+  private def broadcastDelegationTokens(tokens: Array[Byte]) = {
+    logDebug("Sending new tokens to all executors")
```
I'd make this `logInfo` (similar message in YARN code has helped me a lot).

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala:
```diff
@@ -0,0 +1,180 @@
+
+  private def broadcastDelegationTokens(tokens: Array[Byte]) = {
+    logDebug("Sending new tokens to all executors")
+    if (driverEndpoint == null) {
```
Make this a `require` in the constructor?

vanzin reviewed (commented):
Ok, the API looks a lot better. Still a few things to take care of.

Commit:
addressed comments

SparkQA commented:
**[Test build #83660 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83660/testReport)** for PR 19272 at commit [`8df7e37`](https://github.com/apache/spark/commit/8df7e37517a21d5fbaa2c0e7abfa248fd3ff9be3).

Commit:
quick fixup, logging was in the wrong place

SparkQA commented:
**[Test build #83661 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83661/testReport)** for PR 19272 at commit [`45b46ed`](https://github.com/apache/spark/commit/45b46ed6768ea50ddf23063b2a925c2a4794acc7).

SparkQA commented:
**[Test build #83660 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83660/testReport)** for PR 19272 at commit [`8df7e37`](https://github.com/apache/spark/commit/8df7e37517a21d5fbaa2c0e7abfa248fd3ff9be3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83660/
Test PASSed.

SparkQA commented:
**[Test build #83661 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83661/testReport)** for PR 19272 at commit [`45b46ed`](https://github.com/apache/spark/commit/45b46ed6768ea50ddf23063b2a925c2a4794acc7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83661/
Test PASSed.

ArtRand commented:
Hello @vanzin thanks for the continued help with this, anything else needed?

maverick2202 commented:
We want to use this feature and will be great if this can be merged. Any idea which spark release will have it.

ArtRand commented:
Hello @maverick2202, hopefully 2.3 (and maybe back ported?) but that's up to the Committers.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala:
```diff
@@ -0,0 +1,185 @@
+      UserGroupInformation.loginUserFromKeytabAndReturnUGI(principal, secretFile.get)
+    } else {
+      // if the ticket cache is not explicitly defined, use the default
+      if (secretFile.isEmpty) {
```
The point I was trying to make is that you do not need *any* special handling for TGT. The UGI class already does everything you need, you just need to get the current user. It will keep the TGT updated with any changes that happen on disk. You don't need to handle `KRB5CCNAME` anywhere, because UGI should be doing that for you. If it's not, you need to explain why you need this special handling, because the expected behavior is for this to work without you needing to do anything.

So you can simplify this class by *only* handling the principal / keytab case, and just using `UserGroupInformation.getCurrentUser` in the other case. You don't need to keep track of the "mode" or anything else, just whether you're using a principal / keytab pair.

ArtRand replied:
How do the executors get the TGT though?  `KRB5CCNAME` would need to be set in the executor containers as well, right? If it is, I suppose you don't need to broadcast the delegation tokens at all because `runAsSparkUser` takes care of it for you? 

I guess now that https://github.com/apache/spark/pull/19437 has been merged we can disseminate the TGT though Mesos.

vanzin replied:
> How do the executors get the TGT though? 

They don't. That's why you're creating delegation tokens and sending *them* to the executor.

ArtRand replied:
Ah gotcha. I misunderstood your previous comment to mean that you wouldn’t need to renew the tokens when using the ticket cache. I’ll simplify the logic. Thanks.

ArtRand replied:
So `getCurrentUser` actually doesn't work. I believe for the reason mentioned [here](https://github.com/apache/spark/blob/a18d637112b97d2caaca0a8324bdd99086664b24/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMCredentialRenewer.scala#L152). The effect is that the same expired credentials are returned (causing the renewer to loop). 

I understand that using the TGT instead of a keytab is different than the YARN reference implementation, and it's unusual to use it in this case since it expires anyways. Do you think it would be better to avoid the ticket renewal logic all together when using a TGT or to keep the older `UserGroupInformation.getUGIFromTicketCache`-based method?

vanzin replied:
> Do you think it would be better to avoid the ticket renewal logic all together when using a TGT

If `getCurrentUser` does not work, then yes, that's probably the best way forward. Requiring the user to set `KRB5CCNAME` is not really a good way to go about this. Also because in that case the user still has to make sure the TGT is updated through other means.

ArtRand replied:
Agreed, thanks.

Commit:
only renew tokens when using keytab

SparkQA commented:
**[Test build #83863 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83863/testReport)** for PR 19272 at commit [`18d77ff`](https://github.com/apache/spark/commit/18d77ff5a3aa29c9e60538ae87b6d654c229bdfe).

ArtRand commented:
@vanzin PTAL. 
I removed the awkward `mode` parameter from the token manager. Now we only start the renewer thread when using a keytab/principal. The condition is logged appropriately.

SparkQA commented:
**[Test build #83863 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83863/testReport)** for PR 19272 at commit [`18d77ff`](https://github.com/apache/spark/commit/18d77ff5a3aa29c9e60538ae87b6d654c229bdfe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83863/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -140,13 +140,24 @@ class SparkHadoopUtil extends Logging {
+   * Add or overwrite current user's credentials with serialized delegation tokens,
+   * also confirms correct hadoop configuration is set.
+   */
+  def addDelegationTokens(tokens: Array[Byte], sparkConf: SparkConf) {
```
Always forget this class is public. Add `private[spark]`.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -463,6 +474,19 @@ object SparkHadoopUtil {
+   * @param fraction fraction of the time until expiration return
+   * @return Date when the fraction of the time until expiration has passed
+   */
+  def getDateOfNextUpdate(expirationDate: Long, fraction: Double): Long = {
```
Add `private[spark]`.

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala:
```diff
@@ -0,0 +1,157 @@
+  }
+
+  def getTokens(): Array[Byte] = {
+    tokens
```
`tokens` is never updated, so `fetchHadoopDelegationTokens()` will always return the initial set even after it's expired.

ArtRand replied:
Now updated: https://github.com/apache/spark/pull/19272/files#diff-765ac3c4db227cd2c5d796f00794016fR145

vanzin commented on resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala:
```diff
@@ -0,0 +1,157 @@
+  private def getNewDelegationTokens(): Array[Byte] = {
+    logInfo(s"Attempting to login to KDC with principal ${principal}")
+    // Get new delegation tokens by logging in with a new UGI
+    // inspired by AMCredentialRenewer.scala:L174.
```
nit: mentioning line numbers will make this stale very quickly.

Commit:
update tokens, make SparkHadoopUtil methods private

SparkQA commented:
**[Test build #83912 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83912/testReport)** for PR 19272 at commit [`049e4b5`](https://github.com/apache/spark/commit/049e4b554b38f12bd1a2bd6855fcbb6a937fbf99).

SparkQA commented:
**[Test build #83912 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83912/testReport)** for PR 19272 at commit [`049e4b5`](https://github.com/apache/spark/commit/049e4b554b38f12bd1a2bd6855fcbb6a937fbf99).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83912/
Test PASSed.

vanzin commented:
Merging to master.

ArtRand commented:
@vanzin Thanks for the reviews and mentorship!
